src/backend/distributed/commands/function.c

```diff
@@ -157,3 +162,155 @@ EnsureSequentialModeForFunctionDDL(void)
+		 */
+		return false;
+
+	}
+
+	if (!EnableDependencyCreation)
+	{
+		/*
+		 * we are configured to disable object propagation, should not propagate anything
+		 */
+		return false;
+	}
+
+	if (!IsObjectDistributed(address))
+	{
+		/* do not propagate alter function for non-distributed functions */
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * PlanDropFunctionStmt gets called during the planning phase of a DROP FUNCTION statement
+ * and returns a list of DDLJob's that will drop any distributed functions from the
+ * workers.
+ */
+List *
+PlanDropFunctionStmt(DropStmt *stmt, const char *queryString)
+{
+	List *oldObjectWithArgsList = stmt->objects;
+	List *distributedObjectWithArgsList = NIL;
+	List *distributedFunctionAddresses = NIL;
+	ListCell *addressCell = NULL;
+	const char *dropStmtSql = NULL;
+	List *commands = NULL;
+	List *objectAddresses = NIL;
+	ListCell *objectWithArgsListCell = NULL;
+
+	if (creating_extension)
+	{
+		/*
+		 * extensions should be created separately on the workers, types cascading from an
+		 * extension should therefor not be propagated here.
+		 */
+		return NIL;
+	}
+
+	if (!EnableDependencyCreation)
+	{
+		/*
+		 * we are configured to disable object propagation, should not propagate anything
+		 */
+		return NIL;
+	}
+
+
+	/*
+	 * Our statements need to be fully qualified so we can drop them from the right schema
+	 * on the workers
+	 */
+	QualifyTreeNode((Node *) stmt);
+
+	/*
+	 * map all functions to their object addresses and step through them in lock step to
+	 * create two lists with all distributed functions and their addresses.
+	 */
+	objectAddresses = FunctionListToObjectAddresses(oldObjectWithArgsList,
+													stmt->missing_ok);
+	forboth(objectWithArgsListCell, oldObjectWithArgsList, addressCell, objectAddresses)
+	{
+		ObjectAddress *address = (ObjectAddress *) lfirst(addressCell);
+		ObjectWithArgs *func = castNode(ObjectWithArgs, lfirst(objectWithArgsListCell));
+
+		if (!IsObjectDistributed(address))
+		{
+			continue;
+		}
+
+		/* collect information for all distributed functions */
+		distributedFunctionAddresses = lappend(distributedFunctionAddresses, address);
+		distributedObjectWithArgsList = lappend(distributedObjectWithArgsList, func);
+	}
+
+	if (list_length(distributedObjectWithArgsList) <= 0)
+	{
+		/* no distributed functions to drop */
+		return NIL;
+	}
+
+	/*
+	 * managing types can only be done on the coordinator if ddl propagation is on. when
+	 * it is off we will never get here. MX workers don't have a notion of distributed
+	 * types, so we block the call.
+	 */
+	EnsureCoordinator();
+
+	/* remove the entries for the distributed objects on dropping */
+	foreach(addressCell, distributedFunctionAddresses)
+	{
+		ObjectAddress *address = (ObjectAddress *) lfirst(addressCell);
+		UnmarkObjectDistributed(address);
+	}
+
+	stmt->objects = distributedObjectWithArgsList;
+	dropStmtSql = DeparseTreeNode((Node *) stmt);
+	stmt->objects = oldObjectWithArgsList;
```

Comment:
Copying the object before setting the objects seems cleaner. In that case we can have an explicit `distributedDropFunction` vs. `overallDropFunction`, meaning the code clearly documents what it is trying to achieve.